tsqc/pages/frequency_spectra.py

```diff
@@ -0,0 +1,113 @@
+import holoviews as hv
+import numpy as np
+import pandas as pd
+import panel as pn
+
+from .. import config
+
+
+def make_afs_df(ts, span_normalise, afs_type, mode):
+    polarised = False if afs_type == "folded" else True
+
+    afs = ts.allele_frequency_spectrum(
+        span_normalise=span_normalise, windows=None, polarised=polarised, mode=mode
+    )
+    num_samples = ts.num_samples
+    afs_df = pd.DataFrame(
+        {
+            "allele_count": np.arange(1, num_samples + 1),
+            "num_sites": afs[1:],
+        }
+    )
+    return afs_df
+
+
+def make_afs_panel(afs_df, log_bins, mode):
+    len_df = len(afs_df)
+    if log_bins:
+        max_order = int(np.ceil(np.log10(len_df)))
+        bin_edges = np.geomspace(1, 10**max_order, num=max_order + 1).astype(int)
+        labels = [
+            f"({bin_edges[i]}-{bin_edges[i+1]}]" for i in range(len(bin_edges) - 1)
+        ]
+        afs_df["log_bins"] = pd.cut(
+            afs_df["allele_count"],
+            bins=bin_edges,
+            right=True,
+            include_lowest=True,
+            labels=labels,
+        )
+        xdim = "log_bins"
+        xrotation = 0
+    else:
+        num_bins = min(20, int(np.sqrt(len_df)))
+        bin_edges = np.linspace(1, len_df, num_bins).astype(int)
+        labels = [
+            f"({bin_edges[i]}-{bin_edges[i+1]}]" for i in range(len(bin_edges) - 1)
+        ]
+        afs_df["bins"] = pd.cut(
+            afs_df["allele_count"],
+            bins=bin_edges,
+            right=True,
+            include_lowest=True,
+            labels=labels,
+        )
+        xdim = "bins"
+        xrotation = 45
```

Comment:
Could extract the pd.cut out like this to avoid duplication:
```
afs_df[xdim] = pd.cut(
￼            afs_df["allele_count"],
￼            bins=bin_edges,
￼            right=True,
￼            include_lowest=True,
￼            labels=labels,
]```

Reply:
It's not totally obvious to me why there's two different columns ``bins`` and ``log_bins`` here. A comment explaining the rationale would be helpful.